pkg/monitoring/vms/prometheus/prometheus.go

```diff
@@ -0,0 +1,286 @@
+			ch <- mv
+		}
+	}
+	if vmStats.Memory.RSSSet {
+		mv, err := prometheus.NewConstMetric(
+			memoryUsageDesc, prometheus.GaugeValue,
+			float64(vmStats.Memory.RSS),
+			vmStats.Name, "resident",
+		)
+		if err == nil {
+			ch <- mv
+		}
+	}
+}
+
+func updateVcpu(vmStats *stats.DomainStats, ch chan<- prometheus.Metric) {
+	for vcpuId, vcpu := range vmStats.Vcpu {
+		if !vcpu.StateSet || !vcpu.TimeSet {
+			continue
+		}
+		mv, err := prometheus.NewConstMetric(
+			vcpuUsageDesc, prometheus.GaugeValue,
+			float64(vcpu.Time/1000000000),
+			vmStats.Name, fmt.Sprintf("%v", vcpuId), fmt.Sprintf("%v", vcpu.State),
+		)
+		if err != nil {
+			continue
+		}
+		ch <- mv
+	}
+
+}
+
+func updateBlock(vmStats *stats.DomainStats, ch chan<- prometheus.Metric) {
+	for _, block := range vmStats.Block {
+		if !block.NameSet {
+			continue
+		}
+
+		if block.RdReqsSet {
+			mv, err := prometheus.NewConstMetric(
+				storageIopsDesc, prometheus.CounterValue,
+				float64(block.RdReqs),
+				vmStats.Name, block.Name, "read",
+			)
+			if err == nil {
+				ch <- mv
+			}
+		}
+		if block.WrReqsSet {
+			mv, err := prometheus.NewConstMetric(
+				storageIopsDesc, prometheus.CounterValue,
+				float64(block.WrReqs),
+				vmStats.Name, block.Name, "write",
+			)
+			if err == nil {
+				ch <- mv
+			}
+		}
+		if block.FlReqsSet {
+			mv, err := prometheus.NewConstMetric(
+				storageIopsDesc, prometheus.CounterValue,
+				float64(block.FlReqs),
+				vmStats.Name, block.Name, "flush",
+			)
+			if err == nil {
+				ch <- mv
+			}
+		}
+	}
+
+}
+
+func updateNetwork(vmStats *stats.DomainStats, ch chan<- prometheus.Metric) {
+	for _, net := range vmStats.Net {
+		if !net.NameSet {
+			continue
+		}
+		if net.RxBytesSet {
+			mv, err := prometheus.NewConstMetric(
+				networkTrafficDesc, prometheus.CounterValue,
+				float64(net.RxBytes),
+				vmStats.Name, net.Name, "rx",
+			)
+			if err == nil {
+				ch <- mv
+			}
+		}
+		if net.TxBytesSet {
+			mv, err := prometheus.NewConstMetric(
+				networkTrafficDesc, prometheus.CounterValue,
+				float64(net.TxBytes),
+				vmStats.Name, net.Name, "tx",
+			)
+			if err == nil {
+				ch <- mv
+			}
+		}
+	}
+}
+
+func updateVersion(ch chan<- prometheus.Metric) {
+	verinfo := version.Get()
+	ch <- prometheus.MustNewConstMetric(
+		versionDesc, prometheus.GaugeValue,
+		1.0,
+		verinfo.GoVersion, verinfo.GitVersion,
+	)
+}
+
+type Collector struct {
+	virtShareDir string
+}
+
+func SetupCollector(virtShareDir string) *Collector {
+	log.Log.Infof("starting updater: sharedir=%v", virtShareDir)
+	co := &Collector{
+		virtShareDir: virtShareDir,
+	}
+	prometheus.MustRegister(co)
+	return co
+}
+
+func (co Collector) Describe(ch chan<- *prometheus.Desc) {
+	// TODO: Use DescribeByCollect?
+	ch <- versionDesc
+	ch <- storageIopsDesc
+	ch <- vcpuUsageDesc
+	ch <- networkTrafficDesc
+	ch <- memoryUsageDesc
+}
+
+// Note that Collect could be called concurrently
+func (co Collector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
How often is Collect called?

Reply:
Collect is called each time the endpoint is scraped.